packages/curve-ui-kit/src/features/connect-wallet/lib/useWallet.ts

```diff
@@ -31,59 +15,95 @@ const state: {
+  provider,
+  address,
+  ensName,
+}: {
+  chainId: number
+  provider?: Eip1193Provider
+  address: Address
+  ensName?: string | null
+}): Wallet => ({
+  provider,
+  account: { address, ...(ensName && { ensName }) },
+  chainId,
+})
 
-export const useWallet: UseConnectWallet = () => {
-  const [{ wallet: onboardWallet, connecting: onboardConnecting }, onboardConnect, onboardDisconnect] =
-    useOnboardWallet()
-  const [{ wallet: wagmiWallet, connecting: wagmiConnecting }, wagmiConnect, wagmiDisconnect] = useWagmi()
-  const [walletName, setWalletName] = useWalletName()
-  const shouldUseWagmi = useUseWagmi()
+const useWallet = () => {
+  // Dunno why but it was a global and needs to be a global. Can this be a ref?
+  const [showModal, setShowModal] = useGlobalState<boolean>('showConnectModal', false)
+  const closeModal = useCallback(() => setShowModal(false), [setShowModal])
 
-  const { wallet, provider } = useMemo(() => {
-    state.wallet = shouldUseWagmi ? wagmiWallet : onboardWallet && convertOnboardWallet(onboardWallet)
-    state.provider = state.wallet?.provider ? new BrowserProvider(state.wallet.provider) : null
-    return state
-  }, [onboardWallet, wagmiWallet, shouldUseWagmi])
+  const { address } = useAccount()
+  const { data: ensName } = useEnsName({ address })
+  const { data: client } = useConnectorClient()
 
-  const signerAddress = onboardWallet?.accounts[0]?.address ?? wagmiWallet?.account?.address
-  const connect = useMemo(
-    (): ((label?: string) => Promise<Wallet | null>) =>
-      shouldUseWagmi
-        ? wagmiConnect
-        : async (label?: string) => {
-            const [wallet] = await onboardConnect({ ...(label && { autoSelect: { label, disableModals: true } }) })
-            return convertOnboardWallet(wallet)
-          },
-    [onboardConnect, shouldUseWagmi, wagmiConnect],
-  )
+  const wallet =
+    useMemo(
+      () =>
+        client?.transport?.request &&
+        createWallet({
+          chainId: client.chain.id,
+          provider: { request: client.transport.request },
+          address: client.account.address,
+          ensName,
+        }),
+      [client, ensName],
+    ) ?? null
+
+  // important: use the async functions so we can properly handle the promise failures
+  const { connectAsync } = useConnect()
+  const { disconnectAsync } = useDisconnect()
 
-  const disconnect = useMemo(
-    () =>
-      shouldUseWagmi
-        ? wagmiDisconnect
-        : async () => wallet?.label && (await onboardDisconnect({ label: wallet.label })),
-    [onboardDisconnect, shouldUseWagmi, wagmiDisconnect, wallet],
+  /**
+   * Disconnects the wallet from the application.
+   *
+   * This function performs the following steps:
+   * 1. Marks the wallet as manually disconnecting to prevent reconnection attempts
+   * 2. Calls disconnectAsync twice to ensure the useAccount hooks properly reset
+   *    Bro I don't even know why, but if you call it once the useAccount states won't update...
```

Comment:
wtf

Reply:
It seems sometimes when I switch accounts in rabby, the hook doesn't update either. Often it works fine though. Something is wacky